### ISSUE TITLE
`ring` tactic breaks on mathcomp: it requires all ring operations are defined on *syntactically* equal types, not definitionally equal ones.

### ISSUE THREAD

**Issue by Blaisorblade** [coq/coq#11998]
#### Description of the problem

As first noticed by @vlj, the `ring` tactic fails when declaring a ring based on math-comp.
That is fixed by wrapping the needed group operations, such as `+%R`, with definitionally equal ones, such as `addT` below (and using those in the statement): the crucial point is that the type of `addT` arguments is syntactically equal to `R` (more precisely, `GRing.ComRing.sort R`), while `+%R` has a different (but definitionally equal) type (see `check_defeq`).

```
Notation
"+%R" := @GRing.add _ (default interpretation)
GRing.add :
forall (V : GRing.Zmodule.type) (_ : GRing.Zmodule.sort V)
  (_ : GRing.Zmodule.sort V), GRing.Zmodule.sort V
```

```coq
From mathcomp Require Import ssreflect ssrnat eqtype ssrbool ssrnum ssralg.
Require Import Ring.

Import GRing.Theory.
Open Scope ring_scope.

Section foo.
(* Variant of the same problem. *)
(* Variable R: numFieldType. *)
Variable R: comRingType.

Lemma check_defeq: GRing.Zmodule.sort R = GRing.ComRing.sort R. done. Qed.

Locate "+%R".
Check +%R.

Definition addT : R -> R -> R := +%R.
Local Infix "+" := addT.

Definition rt :
  (* Doesn't work.*)
  (* @ring_theory R 0%:R 1%:R +%R *%R (fun a b => a - b) -%R eq. *)
  (* Definitionally equal type, but works. *)
  @ring_theory R 0%:R 1%:R addT *%R (fun a b => a - b) -%R eq.
Proof.
  apply mk_rt.
  - apply add0r.
  - apply addrC.
  - apply addrA.
  - apply mul1r.
  - apply mulrC.
  - apply mulrA.
  - apply mulrDl.
  - by [].
  - apply subrr.
Qed.
Add Ring Ringgg : rt.

Lemma tmp2:
    forall x y:R, x + y = y + x.
Proof. move => x y. ring. Qed.
End foo.
```

The original version of the problem, without all these workarounds, was https://gist.github.com/vlj/caebeb393abf76342a409d871ed39a86/6d24f73884f573521fbac01f40af2caacc0bc14e — that should confirm the issue isn't caused by the workarounds.

#### Coq Version

coq 8.11.dev for me; a release, I expect, for @vlj.

**Comment by vlj** [comment]
Confirming 8.11 release (from opam) on my end

**Comment by herbelin** [comment]
I suspect this comes from the line [there](https://github.com/coq/coq/blob/64b7153a8d48ed5761042b582614ccfa925526d9/plugins/setoid_ring/newring.ml#L341):
```
let field_for_carrier r = Cmap.find r !field_from_carrier 
```
where the search for the domain could be done up to conversion (i.e. using `Reductionops.is_conv` rather than syntactic equality).

I suspect that a quick (local) workaround (not tested) is to use:
```
@ring_theory (GRing.Zmodule.sort R) 0%:R 1%:R +%R *%R (fun a b => a - b) -%R eq.
```

Note that this is not particular specific to mathcomp.

PS: `ring` is an important tactic but my impression is that we are missing maintainers for it...

**Comment by maximedenes** [comment]
> PS: `ring` is an important tactic but my impression is that we are missing maintainers for it...

@thery has been contributing some improvements on ring/field recently. Maybe he'll have an idea about what to do for this.

**Comment by herbelin** [comment]
Ah, good! Sorry about it. I'd become temporarily afraid when searching for "ring" in the issues. Searching for `ring` only in the [title](https://github.com/coq/coq/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+ring+) gives a more reasonable list.

**Comment by thery** [comment]
This is related to 
https://github.com/math-comp/math-comp/issues/401

**Comment by Blaisorblade** [comment]
> I suspect that a quick (local) workaround (not tested) is to use:

You're right, if one updates the rest accordingly (https://gist.github.com/Blaisorblade/990d4ef7418bcbbc0f329dedee5c7b93),
but then `x + y = y + x * 1` doesn't work.

**Comment by herbelin** [comment]
@blaisorblade:
You mean that `*` does not syntactically have the same type `(GRing.Zmodule.sort R)` as `+` have?

@thery:
Are you saying that this is not a `ring` issue but a `mathcomp` issue in the sense that not all operations on the ring structure are syntactically on the same domain? And that this is considered to be already an issue from the point of view of mathcomp itself (independently of the question of ring)?

**Comment by thery** [comment]
> then `x + y = y + x * 1` doesn't work

I guess reification gives some problem here. 1 may not transfered to the one of the ring.

**Comment by thery** [comment]
> Are you saying that this is not a `ring`

No no. It is a ring problem my ref is just that mathcomp people are aware that ring is not working out of the box for mathcomp. I think there are two problems, the first one is the automatic detection of the ring. This is usual the first error you get. The second one is the reification mecanism : be sure that the ring expression that is computed does not miss anything.

**Comment by herbelin** [comment]
I see. Is it in both cases a question of reasoning up to conversion (or at least some conversion)?

**Comment by thery** [comment]
yes so people in mathcomp usually perform sufficient folding/unfolding before calling ring so they know things will go ok.

**Comment by Blaisorblade** [comment]
@herbelin
> You mean that `*` does not syntactically have the same type `(GRing.Zmodule.sort R)` as `+` have?

Indeed not — intuitively, a ring contains something group(or group-like, called Zmodule) and group operators are inherited from the group, so the "group" operators are defined on `GRing.Zmodule.sort R` while the ring operators on `GRing.ComRing.sort R`. And those two types are definitionally but not syntactically equal.

**Comment by herbelin** [comment]
Then, I wonder, will it be enough if we change the line looking for the structure in `newring.ml` (the line I was talking about above), so that it looks up to conversion?

**Comment by herbelin** [comment]
... or are there virtually situations where one may want to have two ring structures on the same domain and still want to discriminate syntactically? In which case, the solution could be to keep the list of syntactic forms used in the declaration and to match only on those (which typically means here keeping the list made of 'GRing.Zmodule.sort R' and `GRing.ComRing.sort R`)?

Or, is there a risk of blowup? If I have a call `tac; try ring` with `ring` supposed to fail on some subgoals, I may prefer that it fails quickly (i.e. syntactically), rather than after having fail to convert?

So, shortly, the first solution (up to conversion) seems mathematically more appealing but the second (up to a list of syntactic forms) may be a better compromise in practise?? I'm missing experience with `ring` to tell.

**Comment by Blaisorblade** [comment]
Many actual ring definitions seem to include some preprocessing in Ltac.

> Or, is there a risk of blowup?

Maybe yes — for typeclasses, people can and do control opacity via `Typeclasses Opaque` and other mechanism, tho I've seldom seen that needed _for types_; OTOH, universe constructions would easily cause that.

For typeclass search, if I don't want a definition to be unfolded, I can export instances of interest and then make it typeclass opaque.
And today `Add Ring` would likely just declare a typeclass instance, just like many specialized commands for setoid rewriting declare instances. (Not sure how hard it'd be to implement).

**Comment by ppedrot** [comment]
Let me chime in about performance issues: it's possible that going for conversion rather than syntactic equality is going to trigger problems, but compared to other issues of the ring tactic, it's unlikely to be the top one. Yet, I'd prefer some opt-in way to call ring to look for rings up to conversion, preferably at the tactic level, rather than hardwiring it in the ML code. Tell me if I am wrong, but assuming your instances are all *reducing* to the ring definition rather than merely convertible to it, we could head-normalize the type with the ring definition set to opaque.

**Comment by herbelin** [comment]
> > then `x + y = y + x * 1` doesn't work
> 
> I guess reification gives some problem here. 1 may not transfered to the one of the ring.

Is it the case that `1` is not syntactically represented by the same structure field when a subterm of `*` or when a subterm of `+`? If yes, that would indeed require to hard-wire some "polysemy" (or reasoning up to conversion) . If not, would declaring two ring instances with well chosen operations (as they appear effectively, i.e. as `Zmodule` or as `comRing`) work:

```coq
Definition rt1 :
 @ring_theory R 0%:R 1%:R +%R *%R (fun a b => a - b) -%R eq.
Add Ring Ringgg1 : rt1.
Definition rt2 :
 @ring_theory (GRing.Zmodule.sort R) 0%:R 1%:R +%R *%R (fun a b => a - b) -%R eq.
Add Ring Ringgg2 : rt2.
```

Incidentally, for the record, @VincentSe and I are considering adding parametric rings for the purpose of reasoning over abstract rings in the [constructive axiomatization](https://github.com/coq/coq/tree/master/theories/Reals/Abstract) of Reals. This would allow to reason over abstract rings without the need for a section.

**Comment by pi8027** [comment]
Hello. Sorry for multi-posting but I implemented [`ring` and `field` tactics for MathComp](https://github.com/math-comp/algebra-tactics) (NB: This is highly experimental and everything is subject to change) with help from Assia and Enrico. These tactics work with any `comRingType` and `fieldType` instances respectively. While there are some performance issues on the reflection side, which I'm going to fix, the reification process that uses conversion seems quite efficient.

**Comment by herbelin** [comment]
Interesting! Can you give more details?

**Comment by pi8027** [comment]
Sure. Roughly speaking, my `ring` tactic checks if the type of the goal unifies with `@eq ?T ?e1 ?e2` first, unifies `?T` with `GRing.ComRing.sort ?ComRing` to obtain the canonical `comRingType` instance of `?T`, and then reifies `?e1` and `?e2` using the instance `?ComRing`.

If the term to reify (such as `?e1` and `?e2`) is *syntactically* an application of a ring operator, say `@GRing.mul ?Ring ?e3 ?e4`, it tries to unify `@GRing.mul ?Ring` with `@GRing.mul (GRing.ComRing.ringType ?ComRing)` to check if the ring operator it found is the ring operator it is looking for. If the unification succeeds, the reification process proceeds to subterms `?e3` and `?e4`.

**Comment by pi8027** [comment]
BTW, unifying `@GRing.mul ?Ring` with `@GRing.mul (GRing.ComRing.ringType ?ComRing)` would be better than unifying `?Ring` with `GRing.ComRing.ringType ?ComRing` for performance reasons, in particular, for concrete ring instances (This was one of the important remarks made by Assia). This optimization may improve the performance of the `zify` tactic, but I actually don't know how it is implemented currently.

**Comment by pi8027** [comment]
For those who did not notice, my [ITP'22 paper](https://doi.org/10.4230/LIPIcs.ITP.2022.29) (particularly Section 3.2) explains my solution to this issue.
